lib/zlib.c

```diff
@@ -0,0 +1,493 @@
+ *
+ * @param {?boolean} [gzip=false]
+ * Add a gzip header if true (creates a gzip-compliant output, otherwise defaults to Zlib)
+ *
+ * @param {?number} [level=Z_DEFAULT_COMPRESSION]
+ * The compression level (0-9).
+ *
+ * @returns {?string}
+ *
+ * @example
+ * // deflate content using default compression
+ * const deflated = deflate(content);
+ *
+ * // deflate content using fastest compression
+ * const deflated = deflate(content, Z_BEST_SPEED);
+ */
+static uc_value_t *
+uc_zlib_deflate(uc_vm_t *vm, size_t nargs)
+{
+	uc_value_t *rv = NULL;
+	uc_value_t *src = uc_fn_arg(0);
+	uc_value_t *gzip = uc_fn_arg(1);
+	uc_value_t *level = uc_fn_arg(2);
+	uc_stringbuf_t *buf = NULL;
+	int ret, lvl = Z_DEFAULT_COMPRESSION;
+	bool gz = false;
+	z_stream strm = {
+		.zalloc = Z_NULL,
+		.zfree = Z_NULL,
+		.opaque = Z_NULL,
+	};
+
+	if (gzip) {
+		if (ucv_type(gzip) != UC_BOOLEAN) {
+			uc_vm_raise_exception(vm, EXCEPTION_TYPE, "Passed gzip flag is not a boolean");
+			goto out;
+		}
+
+		gz = (int)ucv_boolean_get(gzip);
+	}
+
+	if (level) {
+		if (ucv_type(level) != UC_INTEGER) {
+			uc_vm_raise_exception(vm, EXCEPTION_TYPE, "Passed level is not a number");
+			goto out;
+		}
+
+		lvl = (int)ucv_int64_get(level);
+	}
+
+	ret = deflateInit2(&strm, lvl,
+			   Z_DEFLATED,		// only allowed method
+			   gz ? 15+16 : 15,	// 15 Zlib default, +16 for gzip
+			   8,			// default value
+			   Z_DEFAULT_STRATEGY);	// default value
+	if (ret != Z_OK) {
+		uc_vm_raise_exception(vm, EXCEPTION_RUNTIME, "Zlib error: %s", ziniterr(ret));
+		goto out;
+	}
+
+	switch (ucv_type(src)) {
+	case UC_STRING:
+		buf = uc_zlib_def_string(vm, src, &strm);
+		break;
+
+	case UC_RESOURCE:
+	case UC_OBJECT:
+	case UC_ARRAY:
+		buf = uc_zlib_def_object(vm, src, &strm);
+		break;
+
+	default:
+		uc_vm_raise_exception(vm, EXCEPTION_TYPE,
+				      "Passed value is neither a string nor an object");
+		goto out;
+	}
+
+	if (!buf) {
+		if (vm->exception.type == EXCEPTION_NONE)	// do not clobber previous exception
+			uc_vm_raise_exception(vm, EXCEPTION_RUNTIME, "Zlib error: %s", strm.msg);
+		goto out;
+	}
+
+	rv = ucv_stringbuf_finish(buf);
+
+out:
+	(void)deflateEnd(&strm);
+	return rv;
+}
+
+static uc_stringbuf_t *
+uc_zlib_inf_object(uc_vm_t *vm, uc_value_t *obj, z_stream *strm)
+{
+	int ret;
+	bool eof = false;
+	uc_value_t *rfn, *rbuf;
+	uc_stringbuf_t *buf = NULL;
+
+	rfn = ucv_property_get(obj, "read");
+
+	if (!ucv_is_callable(rfn)) {
+		uc_vm_raise_exception(vm, EXCEPTION_TYPE,
+				      "Input object does not implement read() method");
+		return NULL;
+	}
+
+	buf = ucv_stringbuf_new();
+
+	do {
+		rbuf = NULL;
+		uc_vm_stack_push(vm, ucv_get(obj));
+		uc_vm_stack_push(vm, ucv_get(rfn));
+		uc_vm_stack_push(vm, ucv_int64_new(CHUNK));
+
+		if (uc_vm_call(vm, true, 1) != EXCEPTION_NONE)
+			goto fail;
+
+		rbuf = uc_vm_stack_pop(vm);	// read output chunk
+
+		/* we only accept strings */
+		if (ucv_type(rbuf) != UC_STRING) {
```

Comment:
You need to check `rbuf != NULL && ucv_type(rbuf) != UC_STRING` here, as returning `null` from a `read()` method shall be a valid way to signify eof.

Reply:
good catch